lib/_stream_writable.js

```diff
@@ -51,11 +51,20 @@ function WritableState(options, stream) {
   // Note: 0 is a valid value, means that we always return false if
   // the entire buffer is not flushed immediately on write()
   var hwm = options.highWaterMark;
-  var defaultHwm = this.objectMode ? 16 : 16 * 1024;
-  this.highWaterMark = (hwm || hwm === 0) ? hwm : defaultHwm;
+  if (hwm == null) {
+    this.highWaterMark = this.objectMode ? 16 : 16 * 1024;
+  } else {
+    if (typeof hwm !== 'number') {
+      throw new TypeError('"highWaterMark" must be a number');
+    }
 
-  // cast to ints.
-  this.highWaterMark = Math.floor(this.highWaterMark);
+    if (hwm < 0) {
+      throw new RangeError('"highWaterMark" must not be negative');
+    }
+
+    // cast to ints.
+    this.highWaterMark = Math.floor(hwm);
+  }
```

Comment:
I recommend to move this part into a generic function that can be called by readable and writable stream to prevent code duplication. This also keeps it always in line as the limits stay the same in both cases.